src/sentry/integrations/msteams/webhook.py

```diff
@@ -197,10 +225,128 @@ class MsTeamsWebhookEndpoint(Endpoint, MsTeamsWebhookMixin):
+            "external_name": team_name,
+            "installation_type": "team",
+        }
+        return params
+
+    def handle_installation_update_event(self, request: HttpRequest) -> HttpResponse:
+        data = request.data
+        action = data.get("action", None)
+        if action is None or action != "add":
+            logger.info(
+                "sentry.integrations.msteams.webhooks: Action not supported",
+                extra={"request_data": data},
+            )
+            return self.respond({"details": f"{action} is currently not supported"}, status=204)
+
+        try:
+            installation_params = self._get_team_installation_request_data(data=data)
+        except Exception as err:
+            logger.info(
+                "sentry.integrations.msteams.webhooks: Installation param error",
+                exc_info=err,
+                extra={"request_data": data},
+            )
+            return self.respond(
+                {"details": "required request format or keys are missing"}, status=400
+            )
+
+        # sign the params so this can't be forged
+        signed_params = sign(**installation_params)
+
+        # send welcome message to the team
+        preinstall_client = get_preinstall_client(installation_params["service_url"])
+        card = build_team_installation_message(signed_params)
+        preinstall_client.send_card(installation_params["conversation_id"], card)
+
+        return self.respond(status=201)
+
+    def handle_message_event(self, request: HttpRequest) -> HttpResponse:
+        data = request.data
+        conversation = data.get("conversation", {})
+        conversation_type = conversation.get("conversationType")
+
+        # the only message events we care about are those which
+        # are from a user submitting an option on a card, which
+        # will always contain an "payload.actionType" in the data.
+        if data.get("value", {}).get("payload", {}).get("actionType"):
+            # Processing card actions can only occur in the Region silo.
+            if SiloMode.get_current_mode() == SiloMode.CONTROL:
+                return self.respond(status=400)
+            return self.handle_action_submitted(request)
+        elif conversation_type == "channel":
+            return self.handle_channel_message(request)
+
+        return self.handle_personal_message(request)
+
+    def handle_conversation_update_event(self, request: HttpRequest) -> HttpResponse:
+        data = request.data
+        conversation = data.get("conversation", {})
+        conversation_type = conversation.get("conversationType")
+        channel_data = data["channelData"]
+        event = channel_data.get("eventType")
+
+        if event == "teamMemberAdded":
+            return self.handle_team_member_added(request)
+        elif event == "teamMemberRemoved":
+            if SiloMode.get_current_mode() == SiloMode.CONTROL:
+                return self.respond(status=400)
```

Comment:
Why?

Reply:
oh copypaste, cuz proxy. I remembered.

Reply:
this is from existing code, I'm not sure why